File: modules/crash-reporter/android/src/main/java/com/tealium/react/crashreporter/TealiumReactCrashReporter.kt
package com.tealium.react.crashreporter

import android.view.View
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.uimanager.ReactShadowNode
import com.facebook.react.uimanager.ViewManager
import com.tealium.core.Modules
import com.tealium.core.TealiumConfig
import com.tealium.crashreporter.CrashReporter
import com.tealium.crashreporter.truncateCrashReporterStackTraces
import com.tealium.react.OptionalModule
import com.tealium.react.TealiumReact

class TealiumReactCrashReporterPackage : ReactPackage {
    override fun createNativeModules(reactContext: ReactApplicationContext): MutableList<NativeModule> {
        return mutableListOf(TealiumReactCrashReporter(reactContext))
    }

    override fun createViewManagers(reactContext: ReactApplicationContext): MutableList<ViewManager<View, ReactShadowNode<*>>> {
        return mutableListOf()
    }
}

class TealiumReactCrashReporter(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext),
    OptionalModule {

    private var _truncateStackTraces: Boolean = false

    @ReactMethod
    fun initialize(truncateStackTrace: Boolean) {
        _truncateStackTraces = truncateStackTrace
    }

    override fun initialize() {
        super.initialize()
        reactContext.getNativeModule(TealiumReact::class.java)?.registerOptionalModule(this)
    }

    override fun configure(config: TealiumConfig) {
        config.modules.add(Modules.CrashReporter)
        config.truncateCrashReporterStackTraces = _truncateStackTraces
    }

    override fun getName(): String {
        return MODULE_NAME
    }

    companion object {
        const val MODULE_NAME = "TealiumReactCrashReporter"
    }
}

File: modules/crash-reporter/ios/TealiumReactCrashReporter.swift
import Foundation
import tealium_react_native_swift
import TealiumSwift
import TealiumCrashModule

@objc(TealiumReactCrashReporter)
class TealiumReactCrashReporter: NSObject, RCTBridgeModule {
    static func moduleName() -> String! {
        return "TealiumReactCrashReporter"
    }
    let module = TealiumReactCrashReporterModule()

    @objc
    static func requiresMainQueueSetup() -> Bool {
      return false
    }

    override init() {
        super.init()
        TealiumReactNative.registerOptionalModule(module)
    }

    @objc(initialize)
    public func initialize() {
      TealiumReactNative.registerOptionalModule(module)
    }
}

class TealiumReactCrashReporterModule: NSObject, OptionalModule {
  func configure(config: TealiumConfig) {
    config.collectors?.append(Collectors.Crash)
  }
}